src/proposer.ts

```diff
@@ -121,6 +130,246 @@ let s: ReturnType<typeof strings>
 // Initialization flag
 let isInitialized = false
 
+// ~7 days at ~12s blocks
+const APPROX_7D_BLOCKS = 50400
+/**
+ * Validates that a vault ID exists on-chain by checking it is within range
+ * [1, nextVaultId - 1]. Throws if invalid or out of range.
+ */
+export async function validateVaultIdOnChain(vaultId: number): Promise<void> {
+	// Basic sanity
+	if (!Number.isInteger(vaultId) || vaultId < 1) {
+		throw new Error('Invalid vault ID')
+	}
+	// Ensure contracts are initialized
+	if (!vaultTrackerContract) {
+		throw new Error('VaultTracker contract not initialized')
+	}
+	const nextId = await vaultTrackerContract.nextVaultId()
+	const nextIdNumber = Number(nextId)
+	if (!Number.isFinite(nextIdNumber)) {
+		throw new Error('Could not determine nextVaultId from chain')
+	}
+	if (vaultId >= nextIdNumber) {
+		throw new Error('Vault ID does not exist on-chain')
+	}
+}
+
+/**
+ * Computes block range hex strings for Alchemy getAssetTransfers requests,
```

Comment:
Are we getting locked into an Alchemy dependency here or is this something available in generalized libraries for interacting with nodes? Okay if we accept an Alchemy dependency for the moment, just asking.